### PR TITLE
reactors use activejob instead of sidekiq

### PR DESCRIPTION
**Pull request by desheikh**
### Changed
- Reactors now use ActiveJob instead of Sidekiq directly. This allows for more
  flexibility in the future and allows for the use of other job backends.
- The reactor class definition must now inherit from Eventsimple::Reactor and are no longer
  instantiated with the event as an argument. Instead, the event is passed to the `call` method.
  This change is backwards compatible with existing inflight reactor jobs.

### CODE REVIEW

File: lib/eventsimple/configuration.rb [hunk 1]
```diff
@@ -4,13 +4,15 @@ module Eventsimple
...
     attr_accessor :ui_visible_models
 
     def initialize
       @dispatchers = []
       @max_concurrency_retries = 2
       @metadata_klass = 'Eventsimple::Metadata'
+      @active_job_parent_class = 'ActiveJob::Base'
 
       @ui_visible_models = [] # internal use only
     end
```
Comment (AlexRiedler) [inline lib/eventsimple/configuration.rb:18]:
```suggestion
    end
    
    def active_job_parent_class
       @active_job_parent_class.constantize
    end
```

File: lib/eventsimple/configuration.rb [hunk 2]
```diff
@@ -4,13 +4,15 @@ module Eventsimple
...
     attr_writer :metadata_klass
+    attr_accessor :active_job_parent_class
 
     attr_accessor :ui_visible_models
 
     def initialize
       @dispatchers = []
       @max_concurrency_retries = 2
       @metadata_klass = 'Eventsimple::Metadata'
+      @active_job_parent_class = 'ActiveJob::Base'
```
Comment (AlexRiedler) [inline lib/eventsimple/configuration.rb:15]:
```suggestion
```

File: lib/eventsimple/reactor.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+    def perform(event)
+      call(event)
+    end
+
+    around_perform do |job, block|
+      event_global_id = job.arguments.first
+      reactor_class = job.arguments.second
+
+      # For not database based processors like sidekiq, the reactor may trigger before the
```
Comment (cdepaul) [inline lib/eventsimple/reactor.rb:16]:
```suggestion
      # For non database based processors like sidekiq, the reactor may trigger before the
```

File: lib/eventsimple/configuration.rb [hunk 4]
```diff
@@ -2,15 +2,17 @@
 
 module Eventsimple
   class Configuration
-    attr_reader :max_concurrency_retries, :dispatchers
```
Comment (cdepaul) [inline lib/eventsimple/configuration.rb:5]: why remove `:dispatchers`?

Comment (AlexRiedler) [reply lib/eventsimple/configuration.rb:5]: its implemented below on line 36 (https://github.com/wealthsimple/eventsimple/pull/35/files/2b171b444d8846b9087733edf93e9db044c6740c#diff-5c2e10e06f78e6a1037490d78ceaf40b09a5ee2aa2427862a46bdc5a9338beefR36)

File: spec/dummy/config/initializers/eventsimple.rb [hunk 5]
```diff
@@ -3,4 +3,5 @@
   config.dispatchers = %w[
     UserComponent::Dispatcher
   ]
+  config.active_job_parent_klass = 'ApplicationJob'
```
Comment (cdepaul) [inline spec/dummy/config/initializers/eventsimple.rb:6]: Where do we configure that we are using Sidekiq?

Comment (desheikh) [reply spec/dummy/config/initializers/eventsimple.rb:6]: That's up to the base application. We can set `config.active_job.queue_adapter = :sidekiq` for apps that are using sidekiq.

File: lib/eventsimple/engine.rb [hunk 6]
```diff
@@ -33,5 +31,13 @@ class Engine < ::Rails::Engine
         }
       end
     end
+
+    def verify_dispatchers!
+      unless Eventsimple.configuration.dispatchers.all? { |dispatcher|
+        dispatcher.superclass == Eventsimple::Dispatcher
+      }
+        raise ArgumentError, 'dispatchers must inherit from Eventsimple::Dispatcher'
+      end
```
[the remark this answers is not in the document]
Comment (desheikh) [reply lib/eventsimple/engine.rb:40]: 👍 dispatcher isn't an instance for `is_a?` won't work. Changing it to `dispatcher < Eventsimple::Dispatcher`

**Pull request merged**